### PR TITLE
Fix some wibar bugs, polish the API a bit and add some doc.

### PR DESCRIPTION
**Pull request by Elv13**
While I was working on the wallpaper stuff, I found a really bad bug with the wibar. Turns out setting the height has been broken since 4.0. Once upon a time, @psychon let me merge the `awful.placement` code with pretty much "I don't understand this code, but it seems to work". 5 years later, I must admit I no longer understand it either. So fixing it took a bit longer than I hoped... Anyway, it works now.

While I was messing with the placement code anyway (ie, before I forget how it works again), I unbroke the ability to align a wibar, added a margin property (feature request from reddit) and exposed a couple previously hardcoded knobs.

It is still not perfect because the doc claims it is based on `awful.popup`, and it should be, but it is not. Using the popup as the base would allow auto resized wibar to fit the content. But for now this took 3 weeks already, so lets review/merge the part that works.

image

### CODE REVIEW

**Review by Aire-One — CHANGES_REQUESTED** [review]
Thanks for this PR! The wibar margins feature is indeed something that is requested a lot by users. When this will be merged, I'll do another PR to add an example of "dock" style wibar. (Or maybe I'll PR this PR if I do that before it's merged)

> It is still not perfect because the doc claims it is based on awful.popup, and it should be, but it is not. Using the popup as the base would allow auto resized wibar to fit the content. But for now this took 3 weeks already, so lets review/merge the part that works.

Uh, I missed this one... Can we change the @supermodule to wibox, waiting for it to be implemented? It would remove the unwanted API from popup to be part of the wibar doc (This can also be done by overriding the methods with @hide tag).

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:124]: because the `` doesn't copy the code and the `@` does. This example is already used elsewhere in the code.

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:181]: Yeah, your work on the inheritance doc probably gets us closer to that. Back when I originally wrote `06-`, we didn't have the advanced doc generation we have today. That page is actually written in CMakeScript. Porting it to ldoc would be non trivial, but probably a better option. I toyed with the idea of doing it a few time and to list the `@usebeautiful` in a table in the `@beautiful` property doc, but never got very far. As you know, the current state of the template and config.ld are inceasingly hard to extend due to the spagatti of workarounds. Maybe we could add more hacks and build the ldoc template using cmake to combine multiple files, but that's a terrible hack.

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:415]: I am conflicted this this. One way, this brings us closer to QML level of syntax sugar. On the other side, this leaves Lua 5.1 behind. Those meta magic have a big drawback. `pairs()` and `ipairs()` don't work on 5.1 and I don't know how we could fix it beyond monkey-patching `pairs()` and `ipairs()` in both the C and example generation. That might sound like a pretty minor problem, but it actually is worst because LGI and CAPI could not use the `args` if everything was built using `__index` layers.

I added it here because the margins need to be easily enabled/disabled using keybindings and that helps a lot:

```
awful.key({ modkey }, "d", function() s.mywibox.margins.bottom = 0 end),
```

But I am uncertain if this should be globally implemented.

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:618]: the code above depends on the `_private` variables not being nil, so I think it make sense to initialize them to at least the right type before calling the setters

File: tests/examples/screen/template.lua
Comment (Elv13) [reply tests/examples/screen/template.lua:473]: done

**Comment by Elv13** [comment]
Thanks for the quick review! I added your suggested changes.

> Thanks for this PR! The wibar margins feature is indeed something that is requested a lot by users. When this will be merged, I'll do another PR to add an example of "dock" style wibar. (Or maybe I'll PR this PR if I do that before it's merged)

That feature has a loooong history. A basic example would definitively be welcome. I planned on adding on, then it was 2am already when I finished the other images, so I didn't. A few "complete" docks have been implemented over the years, including one by me ([see the left side of the screen](https://raw.githubusercontent.com/Elv13/radical/master/screenshot/all.png) ). I have code to have the auto-hide and display-when-mouse-is-near. I also had the macOS Panther/Tiger rounded corner code, but these days the `gears.shape` API handles that. Our current `.desktop` implementation is too archaic to implement the full "running program dots" like macOS. The [Radical](https://github.com/elv13/radical) implementation works around that by using the GDK/GTK menu code directly rather than re-invent it. But Psychon has always been worried that merging that code leaves us exposed to GTK/GDK introducing a dependency on the window manager and permanently dead-locking awesome after a random GTK update. There is an unmerged PR somewhere with a Lua implementation of the GDK icon/.desktop cache, but that's not really a good path forward either. I think adding real threads (different Lua context + message passing) is the only sane way. Also, as I mentioned a few time recently, since Gnome/KDE moved to CGroups instead of startup notifications, we would need to implement that because not all `.desktop` say which X11 class the program has.

Finally, even if I never published the code and it's old and broken, I did actually implement the zoom and hide dock animations from macOS. It's surprisingly not _that_ hard to do. You just need to make an A1 Cairo surface, paint all icons when use that as shape bounding when there is no compositing (and nothing is required when compositing is there). Then using `mouse::move` to zoom the right area on a sine curve. But performance was mediocre.

Anyway, we can write a very simple documentation example assuming the wibar gets rebased on the popup or a similar auto-resize property is added. With hardcoded width, the example wont be very useful in practice.

> Uh, I missed this one... Can we change the @supermodule to wibox, waiting for it to be implemented? It would remove the unwanted API from popup to be part of the wibar doc (This can also be done by overriding the methods with @hide tag).

I am not sure, maybe we should just fix the code or at least add auto-resize support. This is pretty much mandatory to support macOS style docks. We would need to also figure a way to add "if the length is >= 95% of the screen, start to reduce the height" property, but I am not sure how that would look like?

**Comment by Elv13** [comment]
@Aire-One do you think this is ready to merge?

**Review by Aire-One — DISMISSED** [review]
> @Aire-One do you think this is ready to merge?

Oops sorry, I don't see anything that should hold this PR any longer.

File: lib/awful/wibar.lua
Comment (Aire-One) [inline lib/awful/wibar.lua:393]: Oh yes, I had it wrong in my review... I had to get back at it to see you already fixed it. 👍

**Review by sclu1034 — CHANGES_REQUESTED** [review]
A tip when writing new examples again:

You don't need to put `--DOC_HIDE` into every line individually any more. You can hide whole sections with `--DOC_HIDE_START` and `--DOC_HIDE_END` (e.g. https://github.com/awesomeWM/awesome/blob/master/tests/examples/awful/titlebar/constructor.lua).

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua]: Done, I chose `restrict_workarea`. It still has `workarea` in the name. This name was in the API way before I started contributing and is used in many places, so I think it is consistent to keep it in the name.

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:415]: (sorry for the long delay, as usual)

> I would, instead, propose an API similar to CSS

I would prefer to avoid doing this. Having overlapping properties is quite confusing on its own. I kind of like the "property namespace" the table provides. Plus, the margin concept is used in many places already and use a table. Another concern I have is that it would make the documentation much larger and intimidating. Note that there is precedents to adding properties which are also in table. Some objects have both `:geometry()` and `x`/`y`/`width`/`height`. However that was a "consistency" hack to bring the various wibox/client like API closer to each other (just like I am doing with the wallpaper and dynamic client layout PRs).

I would prefer one of those solution:

 * Remove the meta margin code
 * Keep the meta margin code as-is
 * Refactor to make the meta margin code part of an `gears.object` and begin to integrate it in more places as @Aire-One suggests

I don't really mind which way this goes. There is pros and cons in each. @Aire-One @sclu1034 what's your vote?

> inconsistent with what the property::margins 

Along with the `pairs()` problem, there is some corner case where it will be inconvenient to be perfectly consistent.

**Review by actionless — DISMISSED** [review]
should be alright if it sneak through travis

btw whatsup with Github Actions PR? it seems to need only the approval at this point of time?

**Comment by Elv13** [comment]
> btw whatsup with Github Actions PR? it seems to need only the approval at this point of time?

Yeah, I think its time. Originally, I started paying for Travis because I wanted to get 4.4 out of the door before making major CI changes. But obviously I am beyond late on fixing all the things marked as blockers for 4.4.

**Comment by Elv13** [comment]
Rebased to use GitHub Actions instead of Travis.

@sclu1034 You had some comments above. Do you think this is ok to merge this or do you want some changes?

**Comment by Elv13** [comment]
I plan to merge this PR on Sunday if there is no more comments. This fixes a few real bugs and add much asked-for features.

File: lib/awful/wibar.lua
Comment (Aire-One) [reply lib/awful/wibar.lua:417]: I honestly don't see the first issue raised by @sclu1034 as an unwanted behavior. Yes, if you use the same table reference for multiple objects properties, when you alter this table, all objects are affected. To prevent the side effects from the example, the easiest trick would be to clone the margins tables before providing it to the widget.

Hm... I think we can do the clone in the API, and provide it to the user? `__index = function () return gears.clone(self._private.margins) end`.

File: lib/awful/wibar.lua
Comment (Aire-One) [inline lib/awful/wibar.lua:593]: @Elv13, what are the remaining missing pieces to implement before switching this to use popup?

File: lib/awful/placement.lua
Comment (Aire-One) [inline lib/awful/placement.lua:284]: (This is relevant to all 3 of new locals tables)

Do you think we should expose them as `awful.placement` static fields? I think it could be interesting to prevent duplication of these tables in other modules, like you did here with wibar. It would also "make sure" that the values are correct/consistent across the API.

**Review by Aire-One — COMMENTED** [review]
This is my post merge quick review, just to see if I could catch up something / give my opinion on running discussions.

File: lib/awful/wibar.lua
Comment (Elv13) [reply lib/awful/wibar.lua:593]: I am not sure. Definitively a lot of testing. The size changes due to widget size and the size changes due to screen geometry will most probably fight each other and need some mediation code?

**Pull request merged**
